mypy/typestate.py

```diff
@@ -29,6 +33,31 @@ class TypeState:
     _subtype_caches = {}  # type: ClassVar[Dict[TypeInfo, Set[Tuple[Instance, Instance]]]]
     _subtype_caches_proper = {}  # type: ClassVar[Dict[TypeInfo, Set[Tuple[Instance, Instance]]]]
 
+    # This contains protocol dependencies generated after running a full build,
+    # or after an update. These dependencies are special because:
+    #   * They are a global property of the program; i.e. some dependencies for imported
+    #     classes can be generated in the importing modules.
+    #   * Because of the above, they are serialized separately, after a full run,
+    #     or a full update.
+    # `proto_deps` can be None if after deserialization it turns out that they are
+    # inconsistent with the other cache files (or an error occurred during deserialization).
+    # A blocking error will be generated in this case, since we can't proceed safely.
+    proto_deps = {}  # type: Optional[Dict[str, Set[str]]]
+
+    # Protocols (full names) a given class attempted to implement.
+    # Used to calculate fine grained protocol dependencies and optimize protocol
+    # subtype cache invalidation in fine grained mode.
+    _attempted_protocols = {}  # type: Dict[TypeInfo, Set[str]]
```

Comment:
Should we use the full name instead of a `TypeInfo` reference here as the key? References are harder to reason about since they could refer to per-merged nodes if we are not careful.

Give a concrete example of what's stored here. For example, if we pass a value of type `a.A` to a function expecting something compatible with protocol `p.P`, we'd have a `a.A -> {'p.P', ...}` in the map.

Maybe also discuss the lifetime of this. Does this get flushed after every update?

Reply:
> Should we use the full name instead of a `TypeInfo` reference here as the key? 

I was not sure, but maybe it makes sense to switch to full names, I will give it a try.

> Does this get flushed after every update?

Actually these stay until the daemon shutdown. I think they can be flushed after every update because we capture them in `proto_deps`, I thought it may make sense to keep them for debugging, but maybe it is not necessary if you think they will take too much space.